### PR TITLE
cpp api: Extend the FiniteField related initialization functions

### PR DESCRIPTION
**Pull request by Sarkoxed**
This adds a general way to initialize the finite field and the element of the final field in the C++ API from a string of an arbitrary base (mostly I was thinking about hex strings).

### CODE REVIEW

**Review by alex-ozdemir — APPROVED** [review]
Thank you for this patch.

Let's wait for Aina to review before merging.

**Comment by alex-ozdemir** [comment]
(and, of course, you'll need to update to the latest `main`)

**Comment by aniemetz** [comment]
Also, these new API functions also need to be added for Python and Java. @alex-ozdemir are you going to take care of this?

**Comment by alex-ozdemir** [comment]
> Btw, these new API functions also need to be added for Python and Java. @alex-ozdemir are you going to take care of this?

Yeah, I'll do that in a separate PR (unless you want to do this, @Sarkoxed ).

**Comment by Sarkoxed** [comment]
> > Btw, these new API functions also need to be added for Python and Java. @alex-ozdemir are you going to take care of this?
> 
> Yeah, I'll do that in a separate PR (unless you want to do this, @Sarkoxed ).

I think I'm about to finish the python api functions, however I don't know Java, so I'll have to leave that to you.

**Comment by Sarkoxed** [comment]
Hey, @aniemetz. I've done everything you asked for.

**Comment by Sarkoxed** [comment]
anyone here?...

File: include/cvc5/cvc5.h
Comment (aniemetz) [inline include/cvc5/cvc5.h:3333]:
```suggestion
   * base n.
```

File: include/cvc5/cvc5.h
Comment (aniemetz) [inline include/cvc5/cvc5.h:3335]:
```suggestion
   * @param size The modulus of the field. Must be prime.
```

File: include/cvc5/cvc5.h
Comment (aniemetz) [inline include/cvc5/cvc5.h:3336]:
```suggestion
   * @param base The base of the string representation.
```

File: include/cvc5/cvc5.h
Comment (aniemetz) [inline include/cvc5/cvc5.h:3757]:
```suggestion
   * If `size` is the field size, the constant needs not be in the range [0,size).
```

File: src/api/python/cvc5.pxi
Comment (aniemetz) [inline src/api/python/cvc5.pxi:1515]: Ditto

File: src/api/python/cvc5.pxi
Comment (aniemetz) [inline src/api/python/cvc5.pxi:912]: Why is this an explicit variant?

File: test/unit/api/cpp/solver_black.cpp
Comment (aniemetz) [inline test/unit/api/cpp/solver_black.cpp:515]:
```suggestion
```

File: test/unit/api/cpp/solver_black.cpp
Comment (aniemetz) [inline test/unit/api/cpp/solver_black.cpp:530]: These tests should be added to a `test_mk_finite_field_elem` test in `test/unit/api/python/test_solver.py`. It seems that this test does not exist (@alex-ozdemir looking at you :eyes: ). @Sarkoxed would you mind adding these plus the ones that already existed here to such a new test?

File: test/unit/api/cpp/solver_black.cpp
Comment (aniemetz) [inline test/unit/api/cpp/solver_black.cpp:206]: Same as with `mkFiniteFieldElem` below, @Sarkoxed can you add these and the existing ones to a new test in `test/unit/api/python/test_solver.py`?

File: test/unit/api/python/test_finite_field.py
Comment (aniemetz) [inline test/unit/api/python/test_finite_field.py:57]: This test should also be added for C++.

**Review by aniemetz — CHANGES_REQUESTED** [review]
@Sarkoxed sorry for the delay, been swamped lately.

File: test/unit/api/python/test_finite_field.py
Comment (Sarkoxed) [reply test/unit/api/python/test_finite_field.py:57]: I can't find `test_basic_finite_field` like cpp test, where should I add this?

**Comment by Sarkoxed** [comment]
@aniemetz, Where should I pu `test_basic_finite_field` like cpp test?

File: test/unit/api/python/test_finite_field.py
Comment (aniemetz) [reply test/unit/api/python/test_finite_field.py:57]: Right, you added these two tests, that's why they don't exist for C++ yet. They should be added to https://github.com/cvc5/cvc5/blob/main/test/unit/api/cpp/solver_black.cpp

File: src/api/python/cvc5.pxi
Comment (aniemetz) [inline src/api/python/cvc5.pxi:1517]: Is base 10 really a requirement here? Isn't parameter `base` what determines the base the string is given in?

File: src/api/python/cvc5.pxi
Comment (aniemetz) [inline src/api/python/cvc5.pxi:1519]:
```suggestion
            :param base: The base of the string representation.
```

File: src/api/python/cvc5.pxi
Comment (Sarkoxed) [reply src/api/python/cvc5.pxi:1517]: The default method that is called with one parameter is `Sort mkFiniteFieldElem(const string& s, Sort)` which works only with a base 10 representation. I used mkBitVector creation methods as a reference to this.

File: test/unit/api/python/test_finite_field.py
Comment (Sarkoxed) [reply test/unit/api/python/test_finite_field.py:57]: The first one was not mine, but ok, will be done.

File: src/api/python/cvc5.pxi
Comment (aniemetz) [inline src/api/python/cvc5.pxi:1517]:
```suggestion
            :param value: The value of the element's integer or string (in base
                                      10 representation if no ``base`` given).
```
It seems that this is the behavior you described above? Maybe also add a test for the case when it is not base 10? For example, a hex value?

File: test/unit/api/python/test_finite_field.py
Comment (aniemetz) [reply test/unit/api/python/test_finite_field.py:57]: Ah right, thanks for adding it anyway!

File: src/api/python/cvc5.pxi
Comment (Sarkoxed) [reply src/api/python/cvc5.pxi:1517]: I can simply delete functions like `Sort Solver::mkFiniteFieldSort(const std::string& modulus) const` and keep the `Sort Solver::mkFiniteFieldSort(const std::string& modulus, uint32_t base = 10) const`. Don't know why I didn't suggest it before.

Comment (Sarkoxed) [reply src/api/python/cvc5.pxi:1517]: And also add the tests for the default base

Comment (aniemetz) [reply src/api/python/cvc5.pxi:1517]: Sounds good!

File: include/cvc5/cvc5.h
Comment (aniemetz) [inline include/cvc5/cvc5.h:3452]:
```suggestion
  Sort mkFiniteFieldSort(const std::string& size, uint32_t base = 10) const;
```

File: include/cvc5/cvc5.h
Comment (aniemetz) [inline include/cvc5/cvc5.h:3875]:
```suggestion
  Term mkFiniteFieldElem(const std::string& value, const Sort& sort, uint32_t base = 10) const;
```

File: include/cvc5/cvc5_kind.h
Comment (aniemetz) [inline include/cvc5/cvc5_kind.h:1964]:
```suggestion
    *   - Solver::mkFiniteFieldElem(const std::string&, const Sort&, uint32_t) const
```

File: include/cvc5/cvc5_kind.h
Comment (aniemetz) [inline include/cvc5/cvc5_kind.h:5810]:
```suggestion
   *   - Solver::mkFiniteFieldSort(const std::string&, uint32_t base) const
```

File: src/api/python/cvc5.pxi
Comment (aniemetz) [inline src/api/python/cvc5.pxi:905]: Same here, instead of `*args`, keep `size` and add `base` with a default value

**Review by aniemetz — COMMENTED** [review]
OK, so it seems what we really want is not additional API functions on the C++ level, but an extension with a `base` parameter that has a default value `10`. Similarly, on the Python level, we just want to extend the interface of the functions with a `base` parameter with the same default value. @Sarkoxed can you refactor this accordingly?

**Comment by Sarkoxed** [comment]
Hey, @aniemetz. Improved everything(I guess). Haven't squashed it yet to make diff easier to look through.

File: include/cvc5/cvc5.h
Comment (aniemetz) [inline include/cvc5/cvc5.h:3861]: Switch to match actual order.

File: src/api/python/cvc5.pxi
Comment (aniemetz) [inline src/api/python/cvc5.pxi:912]:
```suggestion
            - ``Sort mkFiniteFieldSort(string size, int base)``
```

File: src/api/python/cvc5.pxi
Comment (aniemetz) [inline src/api/python/cvc5.pxi:1514]:
```suggestion
            - ``Term mkFiniteFieldElem(string value, Sort sort, int base)``
```

File: test/unit/api/python/test_solver.py
Comment (aniemetz) [inline test/unit/api/python/test_solver.py:87]: This is missing a check for `solver.mkFiniteFieldSort` with an `int` and a base other than `10` (which should raise a RuntimeError.

File: test/unit/api/python/test_solver.py
Comment (aniemetz) [inline test/unit/api/python/test_solver.py:418]: This is missing a check for solver.mkFiniteFieldElem with an int value and a base other than 10 (which should raise a RuntimeError.

**Comment by aniemetz** [comment]
@Sarkoxed Building the docs with java bindings fails due to your changes:
```
/home/runner/work/cvc5/cvc5/build-shared/src/api/java/io/github/cvc5/Kind.java:2999: error: unexpected text
   * <li>{@link Solver#mkFiniteFieldElem(String, Sort, int base = 10)} 
         ^
```
```
/home/runner/work/cvc5/cvc5/build-shared/src/api/java/io/github/cvc5/SortKind.java:124: error: unexpected text
   * <li>{@link Solver#mkFiniteFieldSort(String, int base = 10)} 
         ^

**Comment by Sarkoxed** [comment]
@aniemetz Aren't these java issues connected with no java api implementation of these features? I'm kinda lost in code and don't even know where to start to fix that one.

File: cmake/deps-utils/CoCoALib-0.99800-trace.patch
Comment (aniemetz) [inline cmake/deps-utils/CoCoALib-0.99800-trace.patch:10]: Why is this change necessary?

**Review by aniemetz — CHANGES_REQUESTED** [review]
@Sarkoxed sorry for the delay, been traveling. I added the necessary changes to the Java API and several fixes. 3 last things:
- one question about the changes in the cocoa patch
- you need to squash all commits and sign off the squashed commit
- your changes need to be formatted with clang-format (after squashing, e.g., via `git clang-format HEAD~1`)

File: cmake/deps-utils/CoCoALib-0.99800-trace.patch
Comment (alex-ozdemir) [reply cmake/deps-utils/CoCoALib-0.99800-trace.patch:10]: @Sarkoxed : I'm curious about this too.

Comment (Sarkoxed) [reply cmake/deps-utils/CoCoALib-0.99800-trace.patch:10]: Hey @aniemetz, @alex-ozdemir! Sorry for this huge delay.

Regarding the change in the cocoalib patch: I struggled to create a cvc5 for each user on a public server, since the directory `/tmp/CoCoALib-config` has write permissions only for the first user. So, with this patch, it will create a directory for each user. I can remove it from this PR if you don't want this behavior.

**Review by aniemetz — APPROVED** [review]

**Comment by aniemetz** [comment]
@Sarkoxed I had to force push my version of your branch since you lost my commits with your previous changes. But no action item on your side, I'll take care of getting this ready to merge.

**Pull request merged**
